ffi/firewood.go

```diff
@@ -174,13 +179,16 @@ func New(filePath string, conf *Config) (*Database, error) {
 // WARNING: Calling Update with an empty key and nil value will delete the entire database
 // due to prefix deletion semantics.
 //
-// This function is not thread-safe with respect to other calls that reference the latest
-// state of the database, nor any calls that mutate the state of the database.
+// This function conflicts with all other calls that access the latest state of the database,
+// and will lock for the duration of this function.
 func (db *Database) Update(keys, vals [][]byte) (Hash, error) {
 	if db.handle == nil {
 		return EmptyRoot, errDBClosed
 	}
 
+	db.stateLock.Lock()
+	defer db.stateLock.Unlock()
+
```

Comment:
if we check that the handle is null before locking, we need to check again after locking to make sure the handle wasn't set to null while we were waiting for the lock